tensorflow/core/kernels/cholesky_grad.cc

```diff
@@ -0,0 +1,164 @@
+    OP_REQUIRES(context, TensorShapeUtils::IsMatrix(input_tensor_l.shape()),
+                errors::InvalidArgument("In[0] is not a matrix"));
+    OP_REQUIRES(context, TensorShapeUtils::IsMatrix(input_tensor_grad.shape()),
+                errors::InvalidArgument("In[1] is not a matrix"));
+    // Check that input tensors are square.
+    OP_REQUIRES(context,
+                input_tensor_l.dim_size(0) == input_tensor_l.dim_size(1),
+                errors::InvalidArgument("Input matrix must be square."));
+    OP_REQUIRES(context,
+                input_tensor_grad.dim_size(0) == input_tensor_grad.dim_size(1),
+                errors::InvalidArgument("Input matrix must be square."));
+
+    // Check that input tensors are of same size.
+    OP_REQUIRES(context,
+                input_tensor_l.dim_size(0) == input_tensor_grad.dim_size(0),
+                errors::InvalidArgument("Input matrices must be same size."));
+
+    // Create an output tensor
+    Tensor* output_tensor = NULL;
+    OP_REQUIRES_OK(context,
+                   context->allocate_output(0, input_tensor_grad.shape(),
+                   &output_tensor));
+
+    if (output_tensor->NumElements() == 0) {
+      // the output shape is a 0-element matrix, so there is nothing to do.
+      return;
+    }
+    // The next lines are necessary to get Eigen matrix behaviour.
+    const ConstMatrixMap input_matrix_l_full(input_tensor_l.flat<T>().data(),
+                                        input_tensor_l.dim_size(0),
+                                        input_tensor_l.dim_size(1));
+    const ConstMatrixMap input_matrix_grad(input_tensor_grad.flat<T>().data(),
+                                           input_tensor_grad.dim_size(0),
+                                           input_tensor_grad.dim_size(1));
+    MatrixMap output_matrix(output_tensor->template flat<T>().data(),
+                            input_tensor_l.dim_size(0),
+                            input_tensor_l.dim_size(1) );
+
+    // Algorithm only depends on lower triangular half on input_tensor_l.
+    const Matrix input_matrix_l = input_matrix_l_full
+                                  .template triangularView<Eigen::Lower>();
+    // Algorithm only depends on lower triangular half on input_matrix_grad.
+    output_matrix = input_matrix_grad.template triangularView<Eigen::Lower>();
+
+    const int64 kMatrixSize = input_matrix_l.rows();
+    const int64 kMaxBlockSize = 32;
+
+    for (int64 block_end = kMatrixSize;
+               block_end > 0;
+               block_end -= kMaxBlockSize) {
+      const int64 block_begin = std::max(0ll, block_end - kMaxBlockSize);
+      const int64 block_size = block_end - block_begin;
+      const int64 trailing_size = kMatrixSize - block_end;
+      output_matrix.block(block_end, block_begin, trailing_size , block_size)
+                   = input_matrix_l
+                   .block(block_begin, block_begin, block_size, block_size)
+                   .adjoint()
```

Comment:
Hi Alex,

Thanks for the much improved version. I left a few additional comments on
the PR. As Vijay suggests, it might be a good idea to run clang-format on
your .cc and .h files.

Rasmus

On Mon, Mar 28, 2016 at 4:27 PM, Vijay Vasudevan <notifications@github.com>
wrote:

> In tensorflow/core/kernels/cholesky_grad.cc
> <https://github.com/tensorflow/tensorflow/pull/1465#discussion_r57650974>:
>
> > +    // Algorithm only depends on lower triangular half on input_matrix_grad.
> > +    output_matrix = input_matrix_grad.template triangularView<Eigen::Lower>();
> > +
> > +    const int64 kMatrixSize = input_matrix_l.rows();
> > +    const int64 kMaxBlockSize = 32;
> > +
> > +    for (int64 block_end = kMatrixSize;
> > +               block_end > 0;
> > +               block_end -= kMaxBlockSize) {
> > +      const int64 block_begin = std::max(0ll, block_end - kMaxBlockSize);
> > +      const int64 block_size = block_end - block_begin;
> > +      const int64 trailing_size = kMatrixSize - block_end;
> > +      output_matrix.block(block_end, block_begin, trailing_size , block_size)
> > +                   = input_matrix_l
> > +                   .block(block_begin, block_begin, block_size, block_size)
> > +                   .adjoint()
>
> (might be worth running clang-format over the changed lines -- that's what
> we do internally)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/1465/files/37f2bb7b24fd41dec1b5d5d1c2dc1b4c99ef4de7#r57650974>
>